src/buffer/out/Row.hpp

```diff
@@ -20,74 +20,139 @@ Revision History:
 #endif
 
 private:
-    CharRow _charRow;
-    ATTR_ROW _attrRow;
-    LineRendition _lineRendition;
-    til::CoordType _id;
-    til::CoordType _rowWidth;
+    static constexpr uint16_t CharOffsetsTrailer = 0x8000;
+    static constexpr uint16_t CharOffsetsMask = 0x7fff;
+
+    template<typename T>
+    static constexpr uint16_t _clampedUint16(T v) noexcept;
+    template<typename T>
+    constexpr uint16_t _clampedColumn(T v) const noexcept;
+    template<typename T>
+    constexpr uint16_t _clampedColumnInclusive(T v) const noexcept;
+
+    wchar_t _uncheckedChar(size_t off) const noexcept;
+    uint16_t _charSize() const noexcept;
+    uint16_t _uncheckedCharOffset(size_t col) const noexcept;
+    bool _uncheckedIsTrailer(size_t col) const noexcept;
+
+    void _init() noexcept;
+    void _resizeChars(uint16_t colExtEnd, uint16_t chExtBeg, uint16_t chExtEnd, size_t chExtEndNew);
+
+    // These fields are a bit "wasteful", but it makes all this a bit more robust against
+    // programming errors during initial development (which is when this comment was written).
+    // * _chars and _charsHeap are redundant
+    //   If _charsHeap is stored in _chars, we can still infer that
+    //   _chars was allocated on the heap if _chars != _charsBuffer.
+    // * _chars doesn't need a size_t size()
+    //   The size may never exceed an uint16_t anyways.
+    // * _charOffsets doesn't need a size() at all
+    //   The length is already stored in _columns.
+
+    // Most text uses only a single wchar_t per codepoint / grapheme cluster.
+    // That's why TextBuffer allocates a large blob of virtual memory which we can use as
+    // a simplified chars buffer, without having to allocate any additional heap memory.
+    // _charsBuffer fits _columnCount characters at most.
+    wchar_t* _charsBuffer = nullptr;
+    // ...but if this ROW needs to store more than _columnCount characters
+    // then it will allocate a larger string on the heap and store it here.
+    // The capacity of this string on the heap is stored in _chars.size().
+    std::unique_ptr<wchar_t[]> _charsHeap;
+    // _chars either refers to our _charsBuffer or _charsHeap, defaulting to the former.
+    // _chars.size() is NOT the length of the string, but rather its capacity.
+    // _charOffsets[_columnCount] stores the length.
+    std::span<wchar_t> _chars;
+    // _charOffsets accelerates indexing into the above _chars string given a terminal column,
+    // by storing the character index/offset at which a column's text in _chars starts.
+    // It stores 1 more item than this row is wide, allowing it to store the
+    // past-the-end offset, which is thus equal to the length of the string.
+    // For instance given a 4 column ROW containing "abcd" it would store 01234.
+    // Given "a\u732Bd" ("\u732B" is a wide glyph) it would store 01123.
+    // Given "a\uD83D\uDE00d" ("\uD83D\uDE00" is an Emoji) it would store 01134.
+    std::span<uint16_t> _charOffsets;

```

Comment:
Ok, so we have 3 examples:
|||
|--|--|
| `abcd` | 01234 |
| `a\uXXXXd` | 01123 | 
| `a\uXXXX\uXXXXd` | 01134 |

So regular characters are just 1 wide, which means you just increment the next `_charOffsets` by 1.
Wide glyphs are... well... wide... so increment by 2.
Are emojis like extra wide? So increment by 3, but they only take up two cells?

Basically, why `01123` vs `01134`? I don't get why the second to last number is 3 instead of 2.

Reply:
`_charOffsets` doesn't contain how many columns a `wchar_t` is wide, but rather the opposite: The number of `wchar_t`s that a single column holds. In other words, given a column index it tells you the _start_ of the text at that column.

* `01234` = each column contains 1 character, because "abcd" are all narrow glyphs each of which consist of exactly 1 `wchar_t`.
* `01123` = same, but column 1 and 2 refer to the same character (`11`). Because `_charOffsets` contains the same offset twice in a row, we know that this is a wide glyph that is 2 columns wide.
* `01134` = same, but now column 1 and 2 contain the same _two_ characters (`11`). The only difference to the previous example is that this Emoji is a surrogate pair that requires 2 `wchar_t`s worth of storage. But it's still a 2 column wide glyph.

This "offset" approach makes some things a little bit computationally expensive. For instance if your current offset array is `0,1,2,3,4,5,6,7,8,9` and you insert some complex 10 `wchar_t` long Unicode at the start of your row, you end up having to recalculate the offsets of all following columns, so that you end up with `0,10,11,12,13,14,15,16,17,18`. But any alternatives have their own set of flaws. This approach at least makes some other important parts quite simple to implement.

For instance if we need to retrieve the text between column 3 and 7? Easy! You just do `_charOffsets[3]` and `_charOffsets[7]`, do a quick check if you need to extend/shrink the selection so that you don't have cut off wide glyphs, and boom you got the exact range of characters you need to copy from the `_chars` array. Since the extend/shrink is practically only by a single column at most, we end up with a `O(1)` algorithm to retrieve text.